File: 06-DevOps/01-容器化/Seccomp-BPF-系统调用过滤/python/seccomp_demo.py
"""
Seccomp-BPF 程序生成器与教学演示(无 root,生成 BPF 字节流)。

参考资料(已实际阅读,见 README):
  kernel.org userspace-api/seccomp_filter.html
  man7 seccomp(2)             — 6 flags 与 SECCOMP_SET_MODE_FILTER
  man7 prctl(2) PR_SET_SECCOMP

用法:
  python3 seccomp_demo.py dump whitelist "read,write,exit,exit_group" 64
  python3 seccomp_demo.py dump blacklist "rmmod,init_module" 32
  python3 seccomp_demo.py arch-info                         # 列出常见 arch 值
  python3 seccomp_demo.py actions-info                      # 5 种 RET 值十六进制表
"""

import struct
import sys

# syscall 号(按 linux/x86_64 通用)
SYSCALL_NAMES_X86_64 = {
    0: "read", 1: "write", 2: "open", 3: "close", 9: "mmap", 10: "mprotect",
    12: "brk", 13: "rt_sigaction", 14: "rt_sigprocmask", 15: "rt_sigreturn",
    21: "access", 39: "getpid", 56: "clone", 57: "fork", 59: "execve",
    60: "exit", 78: "getdents", 87: "unlink", 217: "getdents64",
    231: "exit_group", 233: "madvise",
}

# SECCOMP_RET_* 来自 kernel.org seccomp_filter.html §"Return values"
SECCOMP_RET_KILL_PROCESS = 0x80000000
SECCOMP_RET_KILL_THREAD  = 0x00000000
SECCOMP_RET_TRAP         = 0x00030000
SECCOMP_RET_ERRNO        = 0x00050000
SECCOMP_RET_LOG          = 0x7FC00000
SECCOMP_RET_USER_NOTIF   = 0x7FC00000  # alias of LOG
SECCOMP_RET_ALLOW        = 0x7FFF0000


def emit(code: int, jt: int = 0, jf: int = 0, k: int = 0) -> bytes:
    """BPF 指令编码 — `struct sock_filter { __u16 code; __u8 jt, jf; __u32 k; }`。

    Linux kernel.seccomp_data.arch 是 32-bit 无符号(AUDIT_ARCH_*),需先把 k 当无符号;但
    SECCOMP_RET_KILL_PROCESS = 0x80000000 也只有 32 位正整数范围。这里 emit 用 "<HBBi" 会
    把超过 INT32_MAX 的值报错,故手动截断 k = k & 0xFFFFFFFF 并以 signed 还原。
    """
    # 注意:k 在 BPF 语境下是 32 位有符号(`int k`),但 uvalue 如 audit arch 是无符号 32-bit
    # 在 SECCOMP_RET_* 范围内(0x80000000~0x7FFF0000)int32 边界上要小心
    k_unsigned = k & 0xFFFFFFFF
    # 用 "I" 编码 + signed int32 截断(对负数 k 也兼容,k & 0x80000000 时按 二补码 还原)
    if k_unsigned & 0x80000000:
        k_s = k_unsigned - 0x100000000
    else:
        k_s = k_unsigned
    return struct.pack("<HBBi", code & 0xFFFF, jt & 0xFF, jf & 0xFF, k_s)


def gen_whitelist(allowed_syscall_names: list[str], arch_aarch64: bool = False) -> list[bytes]:
    """生成"白名单"BPF 程序:仅允许指定 syscall 号;其余 KILL_PROCESS。

    Layout(参考 Mozilla Seccomp BPF Macro):
      1. Load arch,检查
      2. Load syscall nr,A
      3. 对每个允许 nr 做 BPF_JUMP+BPF_JEQ,命中 ret ALLOW
      4. 末尾 ret KILL_PROCESS
    """
    # Linux x86_64 syscall nr 解析(简化;仅给 demo 用)
    nr_by_name = {v: k for k, v in SYSCALL_NAMES_X86_64.items()}
    if arch_aarch64:
        # aarch64 不同号;这里仅做示意
        nr_by_name = {"read": 63, "write": 64, "exit": 93, "exit_group": 94}

    nrs = []
    for name in allowed_syscall_names:
        if name not in nr_by_name:
            raise ValueError(f"unknown syscall in current arch: {name}")
        nrs.append(nr_by_name[name])
    nrs.sort()

    BPF_LD = 0x00
    BPF_W = 0x00
    BPF_ABS = 0x20
    BPF_JMP = 0x05
    BPF_JEQ = 0x10
    BPF_K = 0x00
    BPF_RET = 0x06

    out = []
    # Load arch(AUDIT_ARCH_X86_64 = 0xC000003E 见 linux/audit.h)
    out.append(emit(BPF_LD+BPF_W+BPF_ABS, k=4))           # offsetof(seccomp_data, arch)
    out.append(emit(BPF_JMP+BPF_JEQ+BPF_K, jt=1, jf=0, k=0xC000003E))
    out.append(emit(BPF_RET+BPF_K, k=SECCOMP_RET_KILL_PROCESS))
    # Load syscall nr
    out.append(emit(BPF_LD+BPF_W+BPF_ABS, k=0))           # offsetof(seccomp_data, nr)
    # Linear search:每个 allow 行是 1 JEQ + 1 RET(命中即 ALLOW,跳过下一步)
    for nr in nrs:
        out.append(emit(BPF_JMP+BPF_JEQ+BPF_K, jt=0, jf=1, k=nr))   # jf=1 → 跳到下一个
        out.append(emit(BPF_RET+BPF_K, k=SECCOMP_RET_ALLOW))
    # 不匹配 → KILL
    out.append(emit(BPF_RET+BPF_K, k=SECCOMP_RET_KILL_PROCESS))
    return out


def assemble(bpf_program: list[bytes]) -> bytes:
    """串联 BPF 指令为 flat 字节(模拟 sock_fprog 内存布局)。"""
    return b"".join(bpf_program)


# ---------- 演示 CLI ----------

def dump_action_table():
    print("# SECCOMP_RET_* 值(kernel.org userspace-api/seccomp_filter.html §Return values)")
    table = [
        ("SECCOMP_RET_KILL_PROCESS", SECCOMP_RET_KILL_PROCESS, "整个进程退出,exit=SIGSYS"),
        ("SECCOMP_RET_KILL_THREAD",  SECCOMP_RET_KILL_THREAD,  "仅线程退出"),
        ("SECCOMP_RET_TRAP",         SECCOMP_RET_TRAP,         "发 SIGSYS,handler 可模拟"),
        ("SECCOMP_RET_ERRNO",        SECCOMP_RET_ERRNO,        "返回自定义 errno(低 16 位)"),
        ("SECCOMP_RET_USER_NOTIF",   SECCOMP_RET_USER_NOTIF,   "用户态 fd 收到通知(5.0+)"),
        ("SECCOMP_RET_LOG",          SECCOMP_RET_LOG,          "允许但 audit log(4.14+)"),
        ("SECCOMP_RET_ALLOW",        SECCOMP_RET_ALLOW,        "允许"),
    ]
    for name, v, desc in table:
        print(f"  {name:24s} = {v:#010x}  // {desc}")
    print()
    print("# 多个 filter 同时生效时,取最高优先级:KILL_PROCESS > KILL_THREAD > TRAP > ERRNO > LOG/NOTIF > ALLOW")


def dump_arch_info():
    print("# seccomp_data.arch 常见值(linux/audit.h AUDIT_ARCH_*)")
    arches = [
        ("AUDIT_ARCH_X86_64",  0xC000003E, "64-bit x86 (Linux)"),
        ("AUDIT_ARCH_I386",    0x40000003, "32-bit x86"),
        ("AUDIT_ARCH_AARCH64", 0xC00000B7, "64-bit ARM"),
        ("AUDIT_ARCH_ARM",     0x40000028, "32-bit ARM"),
        ("AUDIT_ARCH_RISCV64", 0xC00000F3, "RISC-V 64-bit"),
        ("AUDIT_ARCH_PPC64",   0xC0000015, "PowerPC 64"),
    ]
    for n, v, d in arches:
        print(f"  {n:18s} = {v:#010x}  // {d}")
    print()
    print("# 同一条 BPF 程序跨 arch 时需重新 encode(由于 syscall 号不同)")
    print("# 正确做法:在 BPF 程序首部校验 arch,再分发到对应分支")


def cmd_dump_whitelist(args):
    if len(args) < 2:
        print("usage: dump whitelist <allowed_syscall_csv> [arch=64]")
        return
    allowed = [s.strip() for s in args[0].split(",") if s.strip()]
    arch_aarch64 = "aarch" in (args[1] if len(args) > 1 else "")
    print(f"# BPF 程序 for 白名单 ({arch_aarch64 and 'aarch64' or 'x86_64'}):")
    print(f"# 允许 syscall: {allowed}")
    print()
    print("/* struct sock_filter filter[] = { */")
    for i, insn_bytes in enumerate(gen_whitelist(allowed, arch_aarch64)):
        # 反解为可读
        code, jt, jf, k = struct.unpack("<HBBi", insn_bytes)
        label = bp_label(code, k)
        print(f"    BPF_INSN({label:<22s}, {jt:#04x}, {jf:#04x}, {k:#010x}),  /* #{i} */")
    print("/* }; */")
    print()
    print("/* struct sock_fprog prog = { .len = N, .filter = filter }; */")


def bp_label(code: int, k: int) -> str:
    """反解 BPF 指令为人读 macro。"""
    LDC = code & 0xFF
    if (code & 0x07) == 0x06:        # BPF_RET
        return f"LD_RET({k:#x})"
    if (code & 0x07) == 0x05:        # BPF_JMP
        return f"JUMP({k:#x})"
    if (code & 0x07) == 0x00:        # BPF_LD
        return f"LOAD({k})"
    return f"RAW({code:#x},{k:#x})"


def main():
    if len(sys.argv) < 2:
        print("usage:")
        print("  python3 seccomp_demo.py dump whitelist \"read,write,exit,exit_group\"")
        print("  python3 seccomp_demo.py arch-info")
        print("  python3 seccomp_demo.py actions-info")
        sys.exit(1)
    head = sys.argv[1]
    if head == "dump" and len(sys.argv) > 2 and sys.argv[2] == "whitelist":
        cmd_dump_whitelist(sys.argv[3:])
    elif head == "arch-info":
        dump_arch_info()
    elif head == "actions-info":
        dump_action_table()
    else:
        print("unknown:", head)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: 06-DevOps/01-容器化/Seccomp-BPF-系统调用过滤/go/seccomp_demo.go
// seccomp_demo.go — Go 版 Seccomp-BPF 演示(无 root + 安装 BPF 演示)
//
// 参考资料:
//   kernel.org userspace-api/seccomp_filter.html
//   man7 seccomp(2)
//   go.syscall.Syscall / SYS_SECCOMP / SYS_PRCTL
//
// 用法:
//   go run seccomp_demo.go dump [whitelist|errno-tcp]
//   go run seccomp_demo.go actions               # 打印 7 种 RET 值
//   go run seccomp_demo.go arch                  # arch 枚举

//go:build linux

package main

import (
	"encoding/binary"
	"fmt"
	"os"
	"runtime"
)

// BPF 指令对齐 linux/filter.h:struct sock_filter { u16 code; u8 jt, jf; u32 k }
type sockFilter struct {
	Code uint16
	Jt   uint8
	Jf   uint8
	K    uint32
}

// SECCOMP_RET_* 来自 kernel.org userspace-api/seccomp_filter.html
const (
	SECCOMP_RET_KILL_PROCESS uint32 = 0x80000000
	SECCOMP_RET_KILL_THREAD  uint32 = 0x00000000
	SECCOMP_RET_TRAP         uint32 = 0x00030000
	SECCOMP_RET_ERRNO        uint32 = 0x00050000
	SECCOMP_RET_USER_NOTIF   uint32 = 0x7FC00000
	SECCOMP_RET_LOG          uint32 = 0x7FC00000
	SECCOMP_RET_ALLOW        uint32 = 0x7FFF0000
)

// 常用 prctl + seccomp 常量(linux/prctl.h + seccomp.h)
const (
	SYS_SECCOMP                 = 317 // x86_64
	SYS_PRCTL                   = 157
	PR_SET_NO_NEW_PRIVS         = 38
	PR_SET_SECCOMP              = 22
	SECCOMP_MODE_FILTER         = 2
	SECCOMP_SET_MODE_FILTER     = 1
	AUDIT_ARCH_X86_64           = 0xC000003E
	AUDIT_ARCH_AARCH64          = 0xC00000B7
	AUDIT_ARCH_I386             = 0x40000003
)

// BPF helper constant
const (
	BPF_LD  = 0x00
	BPF_JMP = 0x05
	BPF_RET = 0x06
	BPF_W   = 0x00
	BPF_ABS = 0x20
	BPF_JEQ = 0x10
	BPF_K   = 0x00
)

// genWhitelist 白名单 BPF 程序(与 Python / C 版同结构)
func genWhitelist(allowed []uint32) []sockFilter {
	out := []sockFilter{
		// Load arch
		{Code: BPF_LD | BPF_W | BPF_ABS, K: 4}, // offsetof(seccomp_data, arch)
		{Code: BPF_JMP | BPF_JEQ | BPF_K, Jt: 1, K: AUDIT_ARCH_X86_64},
		{Code: BPF_RET | BPF_K, K: SECCOMP_RET_KILL_PROCESS},
		// Load nr
		{Code: BPF_LD | BPF_W | BPF_ABS, K: 0},
	}
	for _, nr := range allowed {
		out = append(out, sockFilter{Code: BPF_JMP | BPF_JEQ | BPF_K, Jf: 1, K: nr})
		out = append(out, sockFilter{Code: BPF_RET | BPF_K, K: SECCOMP_RET_ALLOW})
	}
	out = append(out, sockFilter{Code: BPF_RET | BPF_K, K: SECCOMP_RET_KILL_PROCESS})
	return out
}

// printBPF 程序以 Go 数组形式可移植
func printBPF(label string, prog []sockFilter) {
	fmt.Printf("// %s\n// struct sock_filter %s[] = {\n", label, label)
	for i, ins := range prog {
		fmt.Printf("    {Code: %#06x, Jt: %d, Jf: %d, K: %#x},  /* #%d */\n",
			ins.Code, ins.Jt, ins.Jf, ins.K, i)
	}
	fmt.Println("// };")
	// 编码到 bytes(便于贴到 C 文件)
	buf := make([]byte, 0, len(prog)*8)
	for _, ins := range prog {
		buf = binary.LittleEndian.AppendUint16(buf, ins.Code)
		buf = append(buf, ins.Jt, ins.Jf)
		buf = binary.LittleEndian.AppendUint32(buf, ins.K)
	}
	fmt.Printf("// bytes (LE) = %d 字节: % x\n\n", len(buf), buf)
}

func cmdDump(which string) {
	switch which {
	case "whitelist":
		printBPF("whitelist", genWhitelist([]uint32{0, 1, 60, 231, 15}))
	case "errno-tcp":
		// connect (nr=42) → SECCOMP_RET_ERRNO | (ENOSYS=38 << 16)
		prog := []sockFilter{
			{Code: BPF_LD | BPF_W | BPF_ABS, K: 4},
			{Code: BPF_JMP | BPF_JEQ | BPF_K, Jt: 1, K: AUDIT_ARCH_X86_64},
			{Code: BPF_RET | BPF_K, K: SECCOMP_RET_KILL_PROCESS},
			{Code: BPF_LD | BPF_W | BPF_ABS, K: 0},
			{Code: BPF_JMP | BPF_JEQ | BPF_K, Jf: 1, K: 42},
			{Code: BPF_RET | BPF_K, K: SECCOMP_RET_ERRNO | (38 << 16)},
			{Code: BPF_RET | BPF_K, K: SECCOMP_RET_ALLOW},
		}
		printBPF("errno_tcp", prog)
	default:
		fmt.Println("usage: dump whitelist|errno-tcp")
	}
}

func cmdActions() {
	fmt.Println("# SECCOMP_RET_* 值(kernel.org userspace-api/seccomp_filter.html §Return values)")
	fmt.Printf("  SECCOMP_RET_KILL_PROCESS = %#010x\n", SECCOMP_RET_KILL_PROCESS)
	fmt.Printf("  SECCOMP_RET_KILL_THREAD  = %#010x\n", SECCOMP_RET_KILL_THREAD)
	fmt.Printf("  SECCOMP_RET_TRAP         = %#010x\n", SECCOMP_RET_TRAP)
	fmt.Printf("  SECCOMP_RET_ERRNO        = %#010x  // 低 16 位放 errno\n", SECCOMP_RET_ERRNO)
	fmt.Printf("  SECCOMP_RET_USER_NOTIF   = %#010x  // 5.0+\n", SECCOMP_RET_USER_NOTIF)
	fmt.Printf("  SECCOMP_RET_LOG          = %#010x  // 4.14+\n", SECCOMP_RET_LOG)
	fmt.Printf("  SECCOMP_RET_ALLOW        = %#010x\n", SECCOMP_RET_ALLOW)
	fmt.Println()
	fmt.Println("# 优先级(多个 filter 同时):KILL_PROCESS > KILL_THREAD > TRAP > ERRNO > LOG/NOTIF > ALLOW")
}

func cmdArch() {
	fmt.Println("# seccomp_data.arch 常见值(linux/audit.h AUDIT_ARCH_*)")
	fmt.Printf("  AUDIT_ARCH_X86_64    = %#010x\n", AUDIT_ARCH_X86_64)
	fmt.Printf("  AUDIT_ARCH_I386      = %#010x\n", AUDIT_ARCH_I386)
	fmt.Printf("  AUDIT_ARCH_AARCH64   = %#010x\n", AUDIT_ARCH_AARCH64)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("usage: seccomp_demo [dump|actions|arch]")
		os.Exit(1)
	}
	switch os.Args[1] {
	case "dump":
		if len(os.Args) < 3 {
			fmt.Println("usage: dump whitelist|errno-tcp")
			os.Exit(1)
		}
		cmdDump(os.Args[2])
	case "actions":
		cmdActions()
	case "arch":
		cmdArch()
	default:
		if runtime.GOOS != "linux" {
			fmt.Println("(此 demo 设计仅在 Linux 跑实 syscall;非 Linux 平台仅 dump)")
		}
		fmt.Println("unknown:", os.Args[1])
		os.Exit(1)
	}
}
